general/src/processor/selector.rs

```diff
@@ -0,0 +1,214 @@
+use std::fmt;
+use std::str::FromStr;
+
+use failure::Fail;
+use pest::error::Error;
+use pest::iterators::Pair;
+use pest::Parser;
+
+use crate::processor::{ProcessingState, ValueType};
+
+/// Error for invalid selectors
+#[derive(Debug, Fail)]
+pub enum InvalidSelectorError {
+    #[fail(display = "invalid selector: deep wildcard used more than once")]
+    InvalidDeepWildcard,
+
+    #[fail(display = "invalid selector: {}", _0)]
+    ParseError(Error<Rule>),
+
+    #[fail(display = "invalid selector: invalid index")]
+    InvalidIndex,
+
+    #[fail(display = "invalid selector: unknown value")]
+    UnknownType,
+
+    #[fail(display = "parser bug: consumed {} (expected {})", _0, _1)]
+    UnexpectedToken(String, &'static str),
+}
+
+#[derive(Parser)]
+#[grammar = "processor/selector.pest"]
+struct SelectorParser;
```

Comment:
Please wrap this in a module to avoid leakage of generated types:

```rust
mod parser {
    use pest_derive::Parser;

    #[derive(Debug, Parser)]
    #[grammar = "processor/selector.pest"]
    pub struct SelectorParser;
}

use self::parser::{SelectorParser, Rule};
```